Refactor most external libs to be imported targets

nilsdeppe opened the pull request:
## Proposed changes

- Using targets is the modern/correct way to use CMake 3.0 and newer (though this is not well-documented). Some external libs are already implemented as imported targets. This does most of the remaining ones except `Catch`, `Charm`, `jemalloc`, and `tcmalloc`.
- When switch Boost and yamlcpp I changed many of our existing libraries to `PUBLIC`, which has uncovered some cyclic dependencies that I didn't bother resolving here. We will need to go through each library and make sure everything is linked correctly.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Make Blaze an imported target

nilsdeppe committed:
Make Brigand an imported target

nilsdeppe committed:
Make GoogleBenchmark an imported target

nilsdeppe committed:
Add imported targets for NumPy and SciPy

nilsdeppe committed:
Make GNU Scientific Library an imported target

nilsdeppe committed:
No longer include pybind11 everywhere

nilsdeppe committed:
Make LIBXSMM an imported target

nilsdeppe committed:
Make LAPACK an imported target

nilsdeppe committed:
Make Libsharp be an imported target

nilsdeppe committed:
Make Boost an imported library

Since we use boost a lot, basically every library needed to be updated.

nilsdeppe committed:
Remove yaml-cpp from catch-all flags

nilsvu commented on cmake/SetupPch.cmake:
```diff
@@ -68,6 +68,11 @@ if (USE_PCH)
+  target_link_libraries(
+    ${PCH_LIB_NAME}
+    PRIVATE
+    Blaze
```
what is this needed for? I think it would be useful to add a comment here

nilsdeppe replied:
I'm not sure what you mean by `this`. If you mean "Blaze" then well the PCH includes blaze and therefore depends on it, so unclear why we should comment on a dependency in only one place. If you mean the `targed_link_libraries` the PCH depends on Blaze and so to properly work we need to depend on it

nilsvu replied:
yea I wasn't aware `SetupPch.hpp` included blaze, that's fine then

nilsvu commented on src/DataStructures/CMakeLists.txt:
```diff
@@ -17,7 +17,9 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
-  INTERFACE ErrorHandling
+  PUBLIC
+  ErrorHandling
+  Blaze
```
ordering is off

nilsvu commented on src/Utilities/CMakeLists.txt:
```diff
@@ -15,7 +15,9 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
-  INTERFACE ErrorHandling
+  PUBLIC
+  ErrorHandling
+  Blaze
```
ordering is off

nilsvu commented on cmake/SetupNumPy.cmake:
```diff
@@ -11,3 +11,7 @@ file(APPEND
   )
+
+add_library(NumPy INTERFACE IMPORTED)
+set_property(TARGET NumPy PROPERTY
```
can you delete the `spectre_include_directories` on line 8? If no, please add a comment why it's still there

nilsvu commented on cmake/SetupSciPy.cmake:
```diff
@@ -10,3 +10,7 @@ file(APPEND
+
+add_library(SciPy INTERFACE IMPORTED)
+set_property(TARGET SciPy PROPERTY
+  INTERFACE_INCLUDE_DIRECTORIES ${SCIPY_INCLUDE_DIR})
```
since numpy and scipy are not linked with anywhere, what do we need them for? Pybindings should be fine without them. Python dependencies are installed separately anyway

nilsdeppe replied:
I wanted this to be consistent with all our other targets

nilsvu replied:
I suppose my question is: Can we just delete `SetupSciPy.cmake` and `SetupNumPy.cmake`? For Python the'll be installed separately anyway, and I don't think pybind11 needs them within c++.

nilsdeppe replied:
Ah, I see. Let's try it! I would guess we can.

nilsdeppe replied:
We need them for tests that include numpy and scipy

nilsvu replied:
Oh right, sorry I should have remembered that 🤦‍♂️ Then `Test_Pypp` should probably link them

nilsvu replied:
Or we make a new library `PythonTesting` that can be linked with instead of `Test_Pypp` (the latter has test cases for pypp)

nilsdeppe replied:
Agreed, @kidder is currently refactoring our testing code :)

nilsvu commented on cmake/SetupGsl.cmake:
```diff
@@ -22,3 +20,9 @@ file(APPEND
   "GSL Version:  ${GSL_VERSION}\n"
   )
+
+add_library(GnuScientificLibrary INTERFACE IMPORTED)
```
`Gsl`? We use the abbreviation everywhere else in the code

nilsdeppe replied:
My reason is that we already have a `GSL.hpp` which is for a different GSL...

nilsvu replied:
Oh I wasn't aware those are different things until now, have wondered vaguely about this before. It's definitely confusing. Avoiding abbreviation is probably the right thing to do then 👍 perhaps even name `Gsl.hpp` to `GuidelinesSupport.hpp` or something (not in this PR of course)

nilsvu commented on cmake/SetupGsl.cmake:
```diff
@@ -22,3 +20,9 @@ file(APPEND
+set_property(TARGET GnuScientificLibrary PROPERTY
+  INTERFACE_INCLUDE_DIRECTORIES ${GSL_INCLUDE_DIR})
+set_property(TARGET GnuScientificLibrary PROPERTY
+  INTERFACE_LINK_LIBRARIES ${GSL_LIBRARIES})
```
Can you delete lines 6-13 when you replace 27-28 with this:

```cmake
set_property(TARGET GnuScientificLibrary PROPERTY
   INTERFACE_LINK_DIRECTORIES ${GSL_LIBRARY_PATH})
```

(see https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_LINK_DIRECTORIES.html)

Also here a comment would be useful why linking `GSL_LIBRARIES` is not enough

nilsdeppe replied:
There is no `GSL_LIBRARY_PATH` variable https://cmake.org/cmake/help/v3.3/module/FindGSL.html

I'm going to switch to `GSL::gsl`

nilsvu commented on src/NumericalAlgorithms/RootFinding/CMakeLists.txt:
```diff
@@ -12,5 +12,7 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
-  INTERFACE ErrorHandling
+  PUBLIC
+  ErrorHandling
+  GnuScientificLibrary
```
Doesn't `Utilities` need to link with GSL as well? At least for including headers I think

nilsdeppe replied:
Which files in Utilities use GSL?

nilsvu replied:
I was confused about `Gsl.hpp` (see comment above)

nilsvu commented on cmake/SetupLIBXSMM.cmake:
```diff
@@ -12,7 +12,11 @@ file(APPEND
   )
 
-set_property(TARGET Blas
+add_library(LIBXSMM INTERFACE IMPORTED)
```
CamelCase

nilsvu commented on cmake/SetupLIBXSMM.cmake:
```diff
@@ -12,7 +12,11 @@ file(APPEND
   APPEND PROPERTY INTERFACE_LINK_LIBRARIES ${LIBXSMM_LIBRARIES})
+# LIBXSMM falls back to blas, so we need to link against Blas with it as well.
+set_property(TARGET LIBXSMM
+  APPEND PROPERTY INTERFACE_LINK_LIBRARIES ${BLAS_LIBRARIES})
```
Can/should you use the `Blas` target here instead?

nilsvu commented on src/NumericalAlgorithms/LinearAlgebra/CMakeLists.txt:
```diff
@@ -15,4 +15,7 @@ target_link_libraries(
   Blas
   DataStructures
   ErrorHandling
+
```
of these three options of formatting I prefer the first two over the last tbh

```cmake
target_link_libraries(
  ${LIBRARY}
  PUBLIC
  Blas
  DataStructures
  PRIVATE
  Lapack
)
```

```cmake
target_link_libraries(
  ${LIBRARY}
  PUBLIC
    Blas
    DataStructures
  PRIVATE
    Lapack
)
```

```cmake
target_link_libraries(
  ${LIBRARY}
  PUBLIC
  Blas
  DataStructures

  PRIVATE
  Lapack
)
```

nilsdeppe replied:
I find the last to be the easiest by far...

nilsvu commented on cmake/SetupBoost.cmake:
```diff
@@ -3,12 +3,16 @@
+set_property(TARGET Boost
+  APPEND PROPERTY INTERFACE_LINK_LIBRARIES ${Boost_LIBRARIES})
+set_property(TARGET Boost PROPERTY
+  INTERFACE_INCLUDE_DIRECTORIES ${Boost_INCLUDE_DIR})
```
Boost provides a cmake config by default since v1.70.0 or if built with `boost-cmake`. This is automatically loaded by `find_package` if not explicitly disabled (see https://cmake.org/cmake/help/v3.15/module/FindBoost.html#boost-cmake). We should either disable this with `set(Boost_NO_BOOST_CMAKE ON)` (I have to do this manually on my machine because I have a more recent version of boost installed) or require and use the cmake config.

nilsdeppe replied:
There's no way we can rely on CMake 3.15 and boost 1.70 being available, I'll set the disable flag

nilsvu commented on src/NumericalAlgorithms/Spectral/CMakeLists.txt:
```diff
@@ -28,7 +28,12 @@ target_link_libraries(
 
+  INTERFACE
+  # LinearOperators should be PUBLIC but there is a cyclic dependency
+  LinearOperators
```
should't it be the other way round? `LinearOperators` should depend on `Spectral`, not `Spectral` on `LinearOperators`

nilsdeppe replied:
Yes, that's the comment on line 32, and in the comment on the PR I said I didn't fix the cyclic dependencies in this PR

nilsdeppe commented:
Rebased on develop and pushed fixups. The ordering of dependencies in `DataStructures` and `Utilities` resulted in conflicts in several commits so I squashed most of them

nilsvu commented on cmake/SetupLIBXSMM.cmake:
```diff
@@ -12,11 +12,13 @@ file(APPEND
-  APPEND PROPERTY INTERFACE_LINK_LIBRARIES ${BLAS_LIBRARIES})
+add_dependencies(
+  Libxsmm
+  Blas
```
Does `add_dependencies` do what we want? Doesn't it only set the build order, not actually link anything? Your comment on line 20 sounds to me that we need to link Blas into any target that links Libxsmm because functions may use Blas instead. If that's correct then we probably need to
```cmake
set_property(TARGET Libxsmm
   APPEND PROPERTY INTERFACE_LINK_LIBRARIES Blas)
```

nilsvu commented on cmake/SetupGsl.cmake:
```diff
@@ -3,8 +3,6 @@
 find_package(GSL REQUIRED)
 
-spectre_include_directories(${GSL_INCLUDE_DIR})
-list(APPEND SPECTRE_LIBRARIES ${GSL_LIBRARIES})
```
You can delete lines 6-13 as well, now that you're using `GSL::gsl`

nilsdeppe commented:
pushed 2 more fixups for remaining comments

nilsdeppe committed:
Have Pypp depend on SciPy and NumPy

nilsvu commented on cmake/SetupLibsharp.cmake:
```diff
@@ -3,12 +3,17 @@
+
+add_library(Libsharp INTERFACE IMPORTED)
+set_property(TARGET Libsharp PROPERTY
+  INTERFACE_INCLUDE_DIRECTORIES ${LIBSHARP_INCLUDE_DIR})
```
typo: add trailing `S`: `LIBSHARP_INCLUDE_DIRS`

nilsvu commented on src/DataStructures/CMakeLists.txt:
```diff
@@ -17,7 +17,11 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
+  Blaze
+  Boost
+  ErrorHandling
+  Options
```
Missing: `Utilities`

nilsdeppe replied:
If it's okay, the next thing I'm going to do is clean up all the dependencies in the `src` directory, and as part of that this will get fixed. Is that alright?

nilsvu replied:
Unfortunately the branch only compiles on my machine when I add `Utilities` here and then make linking with `DataStructures` public wherever `ConstantExpression.hpp` is being included (so basically everywhere). It's because `ConstantExpression.hpp` includes Blaze..

nilsdeppe replied:
What if you remove `IsVector` from `ConstantExpressions`? I don't see why that include is needed there...

nilsvu replied:
that works 👍 now it's only the libs that include `DenseMatrix` `DenseVector` or `Matrix` that need to link publicly with `DataStructures`. Haven't looked in more detail if that can be avoided, or if it's the same for `DataVector`, `Variables` and friends. I can do more detailed tests tomorrow. I do wonder, however, why you don't have these issues.

nilsdeppe replied:
Okay, I'm hopeful that those are actually correct (or can be corrected in a subsequent PR). I'll add a commit to the beginning that removes the problematic include. My guess is that everything that includes anything from DataStructures needs to link publicly against it, since we use our own data structures in the interfaces to libraries all the time (e.g. DataVector and Tensor)

If you're seeing the issue on macOS, not in the container, then it's just that Apple's linker is doing something different (unclear if more or less correct).

nilsdeppe commented:
@nilsleiffischer I hopefully have addressed your remaining comments. Thanks for the quick and thorough review of this :) Let me know if there's anything else!

nilsdeppe committed:
Remove unused Blaze include in ConstantExpression.hpp

kidder commented on src/NumericalAlgorithms/Interpolation/CMakeLists.txt:
```diff
@@ -32,5 +32,6 @@ target_link_libraries(
   Blas
   Domain
   ErrorHandling
+  GSL::gsl
```
what made this target?

nilsdeppe replied:
CMake's packaged `FindGSL` provides this, at least since CMake 3.3 (I didn't check earlier since that's all we require). Similarly, `FindBoost` exposes `Boost::boost`, etc.

nilsdeppe commented:
@nilsleiffischer I've removed the Blaze include from `ConstantExpressions.hpp` and made all the `DataStructures` libs `PUBLIC`

kidder commented:
@wthrowe you may also want to double-check that changes in this PR will work for your setup...

nilsvu reviewed (dismissed):
I can confirm it works on my machine (macOS with cmake v3.15) 👍

wthrowe commented:
Can you rebase?  I'm trying to test, but I've already hit two failures that I think are unrelated to this PR and fixed on develop.

nilsdeppe commented:
@wthrowe just rebased on develop. Thanks for testing!

wthrowe reviewed (approved):
Works on my machine.  I didn't review all the code changes, but I spot checked a few of the commits and they looked fine.
